File: backend/services/stigmer-server/pkg/domain/artifact/storage/r2_storage.go
package storage

import (
	"bytes"
	"context"
	"fmt"
	"io"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

// R2Storage implements ArtifactStorage using Cloudflare R2 (S3-compatible).
type R2Storage struct {
	client        *s3.Client
	presignClient *s3.PresignClient
	bucket        string
}

// NewR2Storage creates Cloudflare R2-backed artifact storage.
// R2 is S3-compatible, so we use AWS SDK with custom endpoint.
func NewR2Storage(ctx context.Context, cfg Config) (*R2Storage, error) {
	// Validate required config
	if cfg.R2Bucket == "" {
		return nil, fmt.Errorf("R2 bucket name is required")
	}
	if cfg.R2Endpoint == "" {
		return nil, fmt.Errorf("R2 endpoint is required")
	}
	if cfg.R2AccessKeyID == "" {
		return nil, fmt.Errorf("R2 access key ID is required")
	}
	if cfg.R2SecretAccessKey == "" {
		return nil, fmt.Errorf("R2 secret access key is required")
	}

	// Set default region for R2
	region := cfg.R2Region
	if region == "" {
		region = "auto" // R2 typically uses "auto"
	}

	// Create credentials
	creds := credentials.NewStaticCredentialsProvider(
		cfg.R2AccessKeyID,
		cfg.R2SecretAccessKey,
		"",
	)

	// Load AWS config with R2 endpoint
	awsCfg, err := config.LoadDefaultConfig(ctx,
		config.WithRegion(region),
		config.WithCredentialsProvider(creds),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to load AWS config: %w", err)
	}

	// Create S3 client with R2 endpoint
	client := s3.NewFromConfig(awsCfg, func(o *s3.Options) {
		o.BaseEndpoint = aws.String(cfg.R2Endpoint)
		// R2 uses path-style addressing
		o.UsePathStyle = true
	})

	// Create presign client for generating signed URLs
	presignClient := s3.NewPresignClient(client)

	return &R2Storage{
		client:        client,
		presignClient: presignClient,
		bucket:        cfg.R2Bucket,
	}, nil
}

// Upload stores artifact data in R2.
func (r *R2Storage) Upload(ctx context.Context, key string, data []byte, contentType string) error {
	input := &s3.PutObjectInput{
		Bucket: aws.String(r.bucket),
		Key:    aws.String(key),
		Body:   bytes.NewReader(data),
	}

	// Set content type if provided
	if contentType != "" {
		input.ContentType = aws.String(contentType)
	}

	_, err := r.client.PutObject(ctx, input)
	if err != nil {
		return fmt.Errorf("r2 upload failed: %w", err)
	}

	return nil
}

// Download retrieves artifact data from R2.
func (r *R2Storage) Download(ctx context.Context, key string) ([]byte, error) {
	result, err := r.client.GetObject(ctx, &s3.GetObjectInput{
		Bucket: aws.String(r.bucket),
		Key:    aws.String(key),
	})
	if err != nil {
		return nil, fmt.Errorf("r2 download failed: %w", err)
	}
	defer result.Body.Close()

	data, err := io.ReadAll(result.Body)
	if err != nil {
		return nil, fmt.Errorf("r2 read failed: %w", err)
	}

	return data, nil
}

// GetSignedURL generates a presigned URL for downloading the artifact.
// The URL is valid for the specified duration (max 7 days for R2).
//
// When downloadFilename is non-empty, the response Content-Disposition is
// signed into the URL so the browser saves the object as an attachment under
// that name instead of rendering it inline.
func (r *R2Storage) GetSignedURL(ctx context.Context, key string, expiresIn time.Duration, downloadFilename string) (string, error) {
	// R2 has a maximum expiration of 7 days
	maxExpiration := 7 * 24 * time.Hour
	if expiresIn > maxExpiration {
		expiresIn = maxExpiration
	}

	input := &s3.GetObjectInput{
		Bucket: aws.String(r.bucket),
		Key:    aws.String(key),
	}
	if downloadFilename != "" {
		input.ResponseContentDisposition = aws.String(ContentDispositionAttachment(downloadFilename))
	}

	request, err := r.presignClient.PresignGetObject(ctx, input, func(opts *s3.PresignOptions) {
		opts.Expires = expiresIn
	})

	if err != nil {
		return "", fmt.Errorf("r2 presign failed: %w", err)
	}

	return request.URL, nil
}

// Delete removes artifact from R2.
func (r *R2Storage) Delete(ctx context.Context, key string) error {
	_, err := r.client.DeleteObject(ctx, &s3.DeleteObjectInput{
		Bucket: aws.String(r.bucket),
		Key:    aws.String(key),
	})
	if err != nil {
		return fmt.Errorf("r2 delete failed: %w", err)
	}

	return nil
}

// Exists checks if artifact exists in R2.
func (r *R2Storage) Exists(ctx context.Context, key string) (bool, error) {
	_, err := r.client.HeadObject(ctx, &s3.HeadObjectInput{
		Bucket: aws.String(r.bucket),
		Key:    aws.String(key),
	})

	if err != nil {
		// Check if it's a "not found" error
		// AWS SDK v2 doesn't have a simple NotFound checker, so we check the error string
		if isNotFoundError(err) {
			return false, nil
		}
		return false, fmt.Errorf("r2 head failed: %w", err)
	}

	return true, nil
}

// Health checks R2 connectivity.
func (r *R2Storage) Health(ctx context.Context) error {
	_, err := r.client.HeadBucket(ctx, &s3.HeadBucketInput{
		Bucket: aws.String(r.bucket),
	})
	if err != nil {
		return fmt.Errorf("r2 health check failed: %w", err)
	}

	return nil
}

// isNotFoundError checks if the error is a "not found" error.
func isNotFoundError(err error) bool {
	if err == nil {
		return false
	}
	// AWS SDK v2 error checking
	return err.Error() == "NotFound" ||
		err.Error() == "NoSuchKey" ||
		contains(err.Error(), "404") ||
		contains(err.Error(), "not found")
}

// contains is a simple helper to check if a string contains a substring.
func contains(s, substr string) bool {
	return len(s) >= len(substr) && (s == substr ||
		(len(s) > len(substr) && (s[:len(substr)] == substr || s[len(s)-len(substr):] == substr ||
			findSubstring(s, substr))))
}

func findSubstring(s, substr string) bool {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}

File: backend/services/stigmer-server/src/artifactstorage/r2-storage.ts
/**
 * R2 artifact storage — ports pkg/domain/artifact/storage/r2_storage.go:
 * the Cloudflare R2 (S3-compatible) backend, filling the factory arm #17
 * deferred to this sub-project (owner-ratified, approved at the #13 plan
 * gate together with the @aws-sdk dependency).
 *
 * Not conformance-assertable on the local targets (same as Go — the Go
 * tree carries no R2 tests at all); the port adds unit pins for the pieces
 * that have logic: config validation copy, the 7-day presign clamp, the
 * signed Content-Disposition, and the not-found mapping.
 */
import { randomBytes } from "node:crypto";

import {
  DeleteObjectCommand,
  GetObjectCommand,
  HeadBucketCommand,
  HeadObjectCommand,
  PutObjectCommand,
  S3Client,
} from "@aws-sdk/client-s3";
import { getSignedUrl as presignGetObject } from "@aws-sdk/s3-request-presigner";

import type { ArtifactStorage, PresignedUpload } from "./artifact-storage.js";
import {
  ArtifactStorageNotFoundError,
  contentDispositionAttachment,
} from "./artifact-storage.js";

/** Go: "R2 has a maximum expiration of 7 days" — presigns clamp here. */
export const R2_MAX_EXPIRATION_MS = 7 * 24 * 60 * 60 * 1000;

/**
 * Key prefix for presignPut staging blobs (§6b). Pinned as the surface a
 * deployment's bucket lifecycle rule targets: the driver mints under it,
 * operations sweeps it — changing it orphans staged blobs from the sweep.
 */
export const R2_STAGING_PREFIX = "staging/";

/**
 * Go get_content/get_download_url wrap their storage calls in a 30s
 * context timeout. The TS ArtifactStorage interface carries no signal, so
 * the same bound lives one layer down, on the driver's network reads —
 * identical ceiling, applied at the only backend where it can fire.
 */
const R2_CALL_TIMEOUT_MS = 30_000;

export interface R2StorageConfig {
  readonly bucket: string;
  readonly endpoint: string;
  readonly accessKeyId: string;
  readonly secretAccessKey: string;
  /** Usually "auto" for R2 (Go default). */
  readonly region: string;
}

/** Test seam: the two AWS clients, injectable for hermetic unit tests. */
export interface R2Clients {
  readonly client: S3Client;
  readonly presign: typeof presignGetObject;
}

export class R2ArtifactStorage implements ArtifactStorage {
  private readonly client: S3Client;
  private readonly presign: typeof presignGetObject;
  private readonly bucket: string;

  constructor(config: R2StorageConfig, clients?: R2Clients) {
    // Go NewR2Storage's required-config errors, byte-mirrored.
    if (config.bucket === "") {
      throw new Error("R2 bucket name is required");
    }
    if (config.endpoint === "") {
      throw new Error("R2 endpoint is required");
    }
    if (config.accessKeyId === "") {
      throw new Error("R2 access key ID is required");
    }
    if (config.secretAccessKey === "") {
      throw new Error("R2 secret access key is required");
    }

    this.bucket = config.bucket;
    if (clients !== undefined) {
      this.client = clients.client;
      this.presign = clients.presign;
      return;
    }
    this.client = new S3Client({
      region: config.region === "" ? "auto" : config.region,
      endpoint: config.endpoint,
      // R2 uses path-style addressing (Go o.UsePathStyle = true).
      forcePathStyle: true,
      credentials: {
        accessKeyId: config.accessKeyId,
        secretAccessKey: config.secretAccessKey,
      },
    });
    this.presign = presignGetObject;
  }

  async upload(key: string, data: Uint8Array, contentType: string): Promise<void> {
    try {
      await this.client.send(
        new PutObjectCommand({
          Bucket: this.bucket,
          Key: key,
          Body: data,
          ...(contentType !== "" ? { ContentType: contentType } : {}),
        }),
      );
    } catch (error) {
      throw new Error(`r2 upload failed: ${errorText(error)}`, { cause: error });
    }
  }

  async download(key: string): Promise<Uint8Array> {
    let body: Uint8Array;
    try {
      const result = await this.client.send(
        new GetObjectCommand({ Bucket: this.bucket, Key: key }),
        { abortSignal: AbortSignal.timeout(R2_CALL_TIMEOUT_MS) },
      );
      if (result.Body === undefined) {
        throw new Error("empty response body");
      }
      body = await result.Body.transformToByteArray();
    } catch (error) {
      // The typed not-found arm (O5): consumers mapping "missing" onto
      // domain vocabulary branch on the class; every other failure stays
      // the wrapped infrastructure fault it always was. Only the log-line
      // text of the missing-object arm changes — every production caller
      // wraps download faults in a sanitized Internal before the wire.
      if (isNotFoundError(error)) {
        throw new ArtifactStorageNotFoundError(key);
      }
      throw new Error(`r2 download failed: ${errorText(error)}`, { cause: error });
    }
    return body;
  }

  async size(key: string): Promise<number> {
    try {
      const result = await this.client.send(
        new HeadObjectCommand({ Bucket: this.bucket, Key: key }),
      );
      return Number(result.ContentLength ?? 0);
    } catch (error) {
      if (isNotFoundError(error)) {
        throw new ArtifactStorageNotFoundError(key);
      }
      throw new Error(`r2 head failed: ${errorText(error)}`, { cause: error });
    }
  }

  async presignPut(
    declaredSizeBytes: number,
    ttlMs: number,
  ): Promise<PresignedUpload> {
    const clampedMs = Math.min(ttlMs, R2_MAX_EXPIRATION_MS);
    // Random staging key: the URL is the credential (unguessable), and the
    // prefix is the contract surface a bucket lifecycle rule sweeps (§6b —
    // staged uploads the domain never committed must not accrete forever).
    const stagingKey = `${R2_STAGING_PREFIX}${randomBytes(16).toString("hex")}`;
    const command = new PutObjectCommand({
      Bucket: this.bucket,
      Key: stagingKey,
      // Signed into the URL: a PUT whose body disagrees with the declared
      // size fails the signature check — R2's arm of the exact-size
      // contract the local lane enforces by counting bytes.
      ContentLength: declaredSizeBytes,
    });
    try {
      const url = await this.presign(this.client, command, {
        expiresIn: Math.floor(clampedMs / 1000),
      });
      return { url, stagingKey, ttlMs: clampedMs };
    } catch (error) {
      throw new Error(`r2 presign failed: ${errorText(error)}`, { cause: error });
    }
  }

  async getSignedUrl(
    key: string,
    expiresInMs: number,
    downloadFilename: string,
  ): Promise<string> {
    const clampedMs = Math.min(expiresInMs, R2_MAX_EXPIRATION_MS);
    const command = new GetObjectCommand({
      Bucket: this.bucket,
      Key: key,
      // Go signs Content-Disposition into the presigned URL — the browser
      // saves the object under the given name instead of rendering inline.
      ...(downloadFilename !== ""
        ? { ResponseContentDisposition: contentDispositionAttachment(downloadFilename) }
        : {}),
    });
    try {
      return await this.presign(this.client, command, {
        expiresIn: Math.floor(clampedMs / 1000),
      });
    } catch (error) {
      throw new Error(`r2 presign failed: ${errorText(error)}`, { cause: error });
    }
  }

  async delete(key: string): Promise<void> {
    try {
      await this.client.send(
        new DeleteObjectCommand({ Bucket: this.bucket, Key: key }),
      );
    } catch (error) {
      throw new Error(`r2 delete failed: ${errorText(error)}`, { cause: error });
    }
  }

  async exists(key: string): Promise<boolean> {
    try {
      await this.client.send(
        new HeadObjectCommand({ Bucket: this.bucket, Key: key }),
      );
      return true;
    } catch (error) {
      if (isNotFoundError(error)) {
        return false;
      }
      throw new Error(`r2 head failed: ${errorText(error)}`, { cause: error });
    }
  }

  async health(): Promise<void> {
    try {
      await this.client.send(new HeadBucketCommand({ Bucket: this.bucket }));
    } catch (error) {
      throw new Error(`r2 health check failed: ${errorText(error)}`, {
        cause: error,
      });
    }
  }
}

/**
 * Whether the error means "no such object" (Go isNotFoundError, adapted to
 * the SDK's structured errors instead of Go's string matching): the S3
 * NotFound/NoSuchKey names, an HTTP 404, or the legacy text forms.
 */
export function isNotFoundError(error: unknown): boolean {
  if (error === null || error === undefined) {
    return false;
  }
  const name = (error as { name?: string }).name ?? "";
  if (name === "NotFound" || name === "NoSuchKey") {
    return true;
  }
  const statusCode = (error as { $metadata?: { httpStatusCode?: number } })
    .$metadata?.httpStatusCode;
  if (statusCode === 404) {
    return true;
  }
  const message = error instanceof Error ? error.message : String(error);
  return message.includes("404") || message.includes("not found");
}

function errorText(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}
